File: app/src/main/java/com/example/tetrisapp/ui/viewmodel/GameViewModel.java
package com.example.tetrisapp.ui.viewmodel;

import androidx.lifecycle.ViewModel;

import com.example.tetrisapp.model.game.MockPlayfield;
import com.example.tetrisapp.model.game.MockTetris;
import com.example.tetrisapp.model.game.Piece;
import com.example.tetrisapp.model.game.Tetris;
import com.example.tetrisapp.interfaces.PieceConfiguration;
import com.example.tetrisapp.model.game.configuration.PieceConfigurations;
import com.example.tetrisapp.model.local.model.PlayerGameData;
import com.example.tetrisapp.model.local.model.Tetromino;
import com.example.tetrisapp.util.FirebaseTokenUtil;
import com.google.firebase.auth.FirebaseAuth;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.stream.Collectors;

public class GameViewModel extends ViewModel {
    private String idToken = null;

    private final PieceConfiguration configuration = PieceConfigurations.DEFAULT.getConfiguration();
    private final Tetris game = new Tetris(configuration, configuration.getStarterPieces(), configuration.getInitialHistory());

    private final Map<String, PlayerGameData> userGameDataMap = new HashMap<>();

    private final MockTetris mockTetris;
    private final MockPlayfield mockPlayfield = new MockPlayfield();

    private final MockTetris mockTetrisSpectate;
    private final MockPlayfield mockPlayfieldSpectate = new MockPlayfield();

    public GameViewModel() {
        FirebaseTokenUtil.getFirebaseToken(token -> idToken = token);

        mockTetris = new MockTetris();
        mockTetris.setConfiguration(PieceConfigurations.DEFAULT.getConfiguration());
        mockTetris.setPlayfield(mockPlayfield);

        mockTetrisSpectate = new MockTetris();
        mockTetrisSpectate.setConfiguration(PieceConfigurations.DEFAULT.getConfiguration());
        mockTetrisSpectate.setPlayfield(mockPlayfieldSpectate);
    }

    public String updateMockTetris(String currentPlayerUid) {
        List<PlayerGameData> userGameValues = new ArrayList<>(userGameDataMap.values());
        userGameValues.sort(Comparator.comparingInt(o -> o.score));
        userGameValues = userGameValues.stream().filter(PlayerGameData::isPlaying).collect(Collectors.toList());

        PlayerGameData bestScoringPlayer = !userGameValues.isEmpty() ? userGameValues.get(userGameValues.size() - 1) : null;
        if (bestScoringPlayer != null && bestScoringPlayer.getUserId().equals(currentPlayerUid)) {
            bestScoringPlayer = userGameValues.size() >= 2 ? userGameValues.get(userGameValues.size() - 2) : null;
        }

        if (bestScoringPlayer == null) return null;

        // Init current piece
        Piece piece = mockTetris.getConfiguration().get(bestScoringPlayer.tetromino.name).copy();
        piece.setMatrix(bestScoringPlayer.tetromino.matrix);
        piece.setCol(bestScoringPlayer.tetromino.x);
        piece.setRow(bestScoringPlayer.tetromino.y);

        // Init current piece shadow
        Piece pieceShadow = mockTetris.getConfiguration().get(bestScoringPlayer.tetrominoShadow.name).copy();
        pieceShadow.setMatrix(bestScoringPlayer.tetrominoShadow.matrix);
        pieceShadow.setCol(bestScoringPlayer.tetrominoShadow.x);
        pieceShadow.setRow(bestScoringPlayer.tetrominoShadow.y);

        // Set opponent related game views
        mockTetris.setScore(bestScoringPlayer.score);
        mockTetris.setLevel(bestScoringPlayer.level);
        mockTetris.setLines(bestScoringPlayer.lines);
        mockTetris.setTetromino(piece);
        mockTetris.setShadow(pieceShadow);
        mockPlayfield.setState(bestScoringPlayer.playfield);

        return bestScoringPlayer.getUserId();
    }

    public String updateSpectatorMockTetris() {
        List<PlayerGameData> userGameValues = new ArrayList<>(userGameDataMap.values());
        userGameValues.sort(Comparator.comparingInt(o -> o.score));
        userGameValues = userGameValues.stream().filter(PlayerGameData::isPlaying).collect(Collectors.toList());

        PlayerGameData bestScoringPlayer = !userGameValues.isEmpty() ? userGameValues.get(userGameValues.size() - 1) : null;
        if (bestScoringPlayer == null) return null;

        // Init current piece
        Piece piece = mockTetrisSpectate.getConfiguration().get(bestScoringPlayer.tetromino.name).copy();
        piece.setMatrix(bestScoringPlayer.tetromino.matrix);
        piece.setCol(bestScoringPlayer.tetromino.x);
        piece.setRow(bestScoringPlayer.tetromino.y);

        // Init current piece shadow
        Piece pieceShadow = mockTetrisSpectate.getConfiguration().get(bestScoringPlayer.tetrominoShadow.name).copy();
        pieceShadow.setMatrix(bestScoringPlayer.tetrominoShadow.matrix);
        pieceShadow.setCol(bestScoringPlayer.tetrominoShadow.x);
        pieceShadow.setRow(bestScoringPlayer.tetrominoShadow.y);

        // Set opponent related game views
        mockTetrisSpectate.setScore(bestScoringPlayer.score);
        mockTetrisSpectate.setLevel(bestScoringPlayer.level);
        mockTetrisSpectate.setLines(bestScoringPlayer.lines);
        mockTetrisSpectate.setShadow(pieceShadow);
        mockTetrisSpectate.setTetromino(piece);
        mockPlayfieldSpectate.setState(bestScoringPlayer.playfield);

        return bestScoringPlayer.getUserId();
    }

    public int getPlacement() {
        List<PlayerGameData> userGameValues = new ArrayList<>(userGameDataMap.values());
        userGameValues.sort(Comparator.comparingInt(o -> o.score));

        int placement = userGameValues.size() + 1;
        for (int i = 0; i < userGameValues.size(); i++) {
            if (game.getScore() >= userGameValues.get(i).score) {
                placement = userGameValues.size() - i;
            }
        }

        return placement;
    }

    public String getWinnerUid() {
        List<PlayerGameData> userGameValues = new ArrayList<>(userGameDataMap.values());
        userGameValues.sort(Comparator.comparingInt(o -> o.score));

        if (userGameValues.get(userGameValues.size() - 1).score > game.getScore()) {
            return userGameValues.get(userGameValues.size() - 1).getUserId();
        } else {
            return Objects.requireNonNull(FirebaseAuth.getInstance().getCurrentUser()).getUid();
        }
    }

    public PlayerGameData getGameData() {
        return new PlayerGameData(
                game.getScore(),
                game.getLines(),
                game.getLevel(),
                game.getCombo(),
                new Tetromino(
                        game.getCurrentPiece().getName(),
                        game.getCurrentPiece().getMatrix(),
                        game.getCurrentPiece().getCol(),
                        game.getCurrentPiece().getRow()
                ),
                new Tetromino(
                        game.getShadow().getName(),
                        game.getShadow().getMatrix(),
                        game.getShadow().getCol(),
                        game.getShadow().getRow()
                ),
                game.getHeldPiece(),
                game.getTetrominoSequence().toArray(new String[0]),
                game.getPlayfield().getState()
        );
    }

    // Getters

    public Map<String, PlayerGameData> getUserGameDataMap() {
        return userGameDataMap;
    }

    public Tetris getGame() {
        return game;
    }

    public PieceConfiguration getConfiguration() {
        return configuration;
    }

    public String getIdToken() {
        return idToken;
    }

    public MockTetris getMockTetris() {
        return mockTetris;
    }

    public MockPlayfield getMockPlayfield() {
        return mockPlayfield;
    }

    public MockTetris getMockTetrisSpectate() {
        return mockTetrisSpectate;
    }

    public MockPlayfield getMockPlayfieldSpectate() {
        return mockPlayfieldSpectate;
    }
}

File: app/src/main/java/com/example/tetrisapp/ui/viewmodel/GameViewModel.kt
package com.example.tetrisapp.ui.viewmodel

import androidx.lifecycle.ViewModel
import com.example.tetrisapp.interfaces.PieceConfiguration
import com.example.tetrisapp.model.game.MockPlayfield
import com.example.tetrisapp.model.game.MockTetris
import com.example.tetrisapp.model.game.Piece
import com.example.tetrisapp.model.game.Tetris
import com.example.tetrisapp.model.game.configuration.PieceConfigurations
import com.example.tetrisapp.model.local.model.PlayerGameData
import com.example.tetrisapp.model.local.model.Tetromino
import com.example.tetrisapp.util.FirebaseTokenUtil
import com.example.tetrisapp.util.PusherUtil.getUserInfo
import com.pusher.client.channel.PresenceChannel
import java.util.*
import java.util.concurrent.Executors
import java.util.concurrent.Future
import java.util.concurrent.ScheduledFuture

class GameViewModel : ViewModel() {
    var token: String? = null
    var timer: Int = 0
    var multiplayerTimer: Int = 0
    lateinit var configuration: PieceConfiguration
    lateinit var game: Tetris
    var countdown = 0
    var countdownRemaining = 0

    val executor = Executors.newSingleThreadScheduledExecutor()
    var futureMoveRight: ScheduledFuture<*>? = null
    var futureMoveLeft: ScheduledFuture<*>? = null
    var countdownFuture: Future<*>? = null
    var timerFuture: ScheduledFuture<*>? = null

    val userGameDataMap: MutableMap<String, PlayerGameData> = HashMap()
    val mockTetris = MockTetris()
    val mockPlayfield = MockPlayfield()
    val mockTetrisSpectate = MockTetris()
    val mockPlayfieldSpectate = MockPlayfield()

    var placement = 1

    init {
        FirebaseTokenUtil.getFirebaseToken { token: String? -> this.token = token }
        mockTetris.playfield = mockPlayfield
        mockTetrisSpectate.playfield = mockPlayfieldSpectate
    }

    fun updateMockTetris(channel: PresenceChannel, currentPlayerUid: String): String? {
        val userGameValues: MutableList<PlayerGameData> = userGameDataMap.values.toMutableList()

        userGameValues.sortBy { data -> -data.score }
        userGameValues.filter { data -> data.isPlaying }

        var bestScoringPlayer = userGameValues.lastOrNull()
        if (bestScoringPlayer?.userId == currentPlayerUid) {
            bestScoringPlayer = if (userGameValues.size >= 2) userGameValues[userGameValues.size - 2] else null
        }

        if (bestScoringPlayer == null) return null
        val configurationName = getUserInfo(channel, bestScoringPlayer.userId)?.configuration ?: PieceConfigurations.DEFAULT.name
        val configuration = PieceConfigurations.valueOf(configurationName).configuration

        mockTetris.configuration = configuration

        // Init current piece
        val piece = mockTetris.configuration[bestScoringPlayer.tetromino.name].copy()
        piece.matrix = bestScoringPlayer.tetromino.matrix
        piece.col = bestScoringPlayer.tetromino.x
        piece.row = bestScoringPlayer.tetromino.y

        // Init current piece shadow
        val pieceShadow = mockTetris.configuration[bestScoringPlayer.tetrominoShadow.name].copy()
        pieceShadow.matrix = bestScoringPlayer.tetrominoShadow.matrix
        pieceShadow.col = bestScoringPlayer.tetrominoShadow.x
        pieceShadow.row = bestScoringPlayer.tetrominoShadow.y

        // Set opponent related game views
        mockTetris.score = bestScoringPlayer.score
        mockTetris.level = bestScoringPlayer.level
        mockTetris.lines = bestScoringPlayer.lines
        mockTetris.currentPiece = piece
        mockTetris.shadow = pieceShadow
        mockPlayfield.state = bestScoringPlayer.playfield
        return bestScoringPlayer.userId
    }

    fun updateSpectatorMockTetris(channel: PresenceChannel): String? {
        val userGameValues: MutableList<PlayerGameData> = userGameDataMap.values.toMutableList()

        userGameValues.sortBy { data -> -data.score }
        userGameValues.filter { data -> data.isPlaying }

        val bestScoringPlayer = userGameValues.lastOrNull() ?: return null

        val configurationName = getUserInfo(channel, bestScoringPlayer.userId)?.configuration ?: PieceConfigurations.DEFAULT.name
        val configuration = PieceConfigurations.valueOf(configurationName).configuration

        mockTetrisSpectate.configuration = configuration

        // Init current piece
        val piece = mockTetrisSpectate.configuration[bestScoringPlayer.tetromino.name].copy()
        piece.matrix = bestScoringPlayer.tetromino.matrix
        piece.col = bestScoringPlayer.tetromino.x
        piece.row = bestScoringPlayer.tetromino.y

        // Init current piece shadow
        val pieceShadow =
            mockTetrisSpectate.configuration[bestScoringPlayer.tetrominoShadow.name].copy()
        pieceShadow.matrix = bestScoringPlayer.tetrominoShadow.matrix
        pieceShadow.col = bestScoringPlayer.tetrominoShadow.x
        pieceShadow.row = bestScoringPlayer.tetrominoShadow.y

        // Set opponent related game views
        mockTetrisSpectate.score = bestScoringPlayer.score
        mockTetrisSpectate.level = bestScoringPlayer.level
        mockTetrisSpectate.lines = bestScoringPlayer.lines
        mockTetrisSpectate.shadow = pieceShadow
        mockTetrisSpectate.currentPiece = piece
        mockPlayfieldSpectate.state = bestScoringPlayer.playfield
        return bestScoringPlayer.userId
    }

    fun getPlacement(gameMode: String): Int {
        return if (gameMode == "battleRoyale") {
            placement
        } else {
            val userGameValues: MutableList<PlayerGameData> = userGameDataMap.values.toMutableList()
            userGameValues.sortBy { data -> -data.score }

            var placement = userGameValues.size + 1
            for (i in userGameValues.indices) {
                if (game.score >= userGameValues[i].score) {
                    placement = i + 1
                    break
                }
            }
            placement
        }
    }

    fun countPlaying(): Int {
        return userGameDataMap.values.count { it.isPlaying } + if (game.isGameOver) 0 else 1
    }

    fun getGameData(userId: String?): PlayerGameData {
        var placement: Int? = null
        if (game.isGameOver) {
            placement = userGameDataMap.values.stream()
                .reduce(0, { x: Int, y: PlayerGameData ->
                    x + if (y.isPlaying) 1 else 0 })
                { a: Int, b: Int ->
                    a + b
                } + 1
        }

        return PlayerGameData(
            userId,
            game.score,
            game.lines,
            game.level,
            game.combo,
            pieceToTetromino(game.currentPiece!!),
            pieceToTetromino(game.shadow!!),
            game.heldPiece,
            game.tetrominoSequence.toTypedArray(),
            game.playfield.state,
            !game.isGameOver,
            placement
        )
    }

    private fun pieceToTetromino(piece: Piece): Tetromino {
        return Tetromino(
            piece.name,
            piece.matrix,
            piece.col,
            piece.row
        )
    }

    companion object {
        const val TAG = "GameViewModel"
    }
}